### PR TITLE
Type and symbol resolving

### PR DESCRIPTION
**Pull request by 3flex**
This PR adds type and symbol resolving for Kotlin code that targets the JVM, seeking early feedback. It's a WIP until at a minimum the following tests/features are added:

- [x] Make the functionality optional - it's slow, so want to disable by default unless a value passed to `--classpath` parameter
- [ ] Tests for the new functionality in a mixed Java/Kotlin codebase
- [x] Tests for the new `UselessCallOnNotNull` rule
- [x] Sharing compiler errors/warnings with the user (they are not currently shown, so user has to know that their code compiles before running detekt, otherwise there'll be an error)

Kotlin/JS and Kotlin/Native will not be in scope, so naturally won't look at MPP support either. Will tackle Gradle plugin integration in future but not as part of this PR.

To test, enable the `UselessCallOnNotNull` rule in the config file, add the following test file with a `kts` extension, then run with `--input path/to/testfile.kts --classpath path/to/kotlin-stdlib.jar --config path/to/config.yml`.
```kotlin
val nextTest = mutableListOf<String>().orEmpty()
```
The new rule will resolve the `orEmpty()` call to `kotlin.collections.orEmpty` which will trigger a detekt issue because the `orEmpty()` call is never needed on a not-null collection.

This is an updated version of #485, rebased and with a simplified implementation (and the intention to get it merged this time! :smile:).

### CODE REVIEW

**Comment by schalkms** [comment]
Hey! Thanks for the tremendous work you spent on this one. 🥇 
I'll review it very soon.

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/Configurations.kt [hunk 1]
```diff
@@ -23,6 +23,10 @@ fun CliArgs.createPlugins(): List<Path> = plugins.letIfNonEmpty {
     MultipleExistingPathConverter().convert(this)
 }
 
+fun CliArgs.createClasspath(): List<String> = classpath.letIfNonEmpty {
+    this.split(";")
```
Comment (arturbosch) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/Configurations.kt]: nit: on the same line it would look better :)

**Comment by arturbosch** [comment]
Yeah, thanks you very much for picking this up.
Looks good, will clone your branch and see how the performance is ^^

**Comment by 3flex** [comment]
Performance will be poor. It's effectively compiling the application (or at least all the Kotlin code) before running the rule sets.

Could probably do something creative in the intellij plugin which allows hooking into the lazy resolve capabilities built into the IDE, which is how the Kotlin plugin runs it's checks.

But to run on the CLI in a self contained tool this is the only way I'm aware of. The value of the implemented rules will have to outweigh the performance cost.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt [hunk 2]
```diff
@@ -24,12 +25,18 @@ abstract class BaseRule(
...
 
     open val ruleId: RuleId = javaClass.simpleName
+    var bindingContext: BindingContext = BindingContext.EMPTY
 
     /**
      * Before starting visiting kotlin elements, a check is performed if this rule should be triggered.
      * Pre- and post-visit-hooks are executed before/after the visiting process.
+     * BindingContext holds the result of the semantic analysis of the source code by the Kotlin compiler. Rules that
+     * rely on symbols and types being resolved can use the BindingContext for this analysis. Note that
+     * BindingContext will have the value BindingContext.EMPTY unless detekt is provided the correct classpath while
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt]: is providing

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+import org.jetbrains.kotlin.types.TypeUtils
+
+/**
+ * TBC
+ *
+ * @author Matthew Haughton
+ */
+class UselessCallOnNotNull(config: Config = Config.empty) : Rule(config) {
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: I'll review this rule when it's finished.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt (hunk 2 above)
Comment (3flex) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt]: It's trying to say that the classpath must be provided to detekt, which then provides the BindingContext to the rule.

I'll reword it though to try and make that clearer.

**Comment by 3flex** [comment]
I'm pretty happy where this is at - appreciate a closer review.

The main new feature since original submission is the fact that errors are shown when the input is being evaluated and the code cannot compile for some reason (and would thus not provide valid results when attempting type and symbol resolution). This will help when not all required dependencies are included on the classpath input.

Mixed Kotlin/Java inputs will also work if both `.kt` and `.java` files (or paths containing them) are provided to the `--input` flag.

Test for the new rule is there, but how to go about efficiently testing the additions in detekt-core and detekt-api are a bit beyond me. When integration with Gradle is added there'll be some good testing done from there so maybe that will be enough?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import org.jetbrains.kotlin.psi.KtQualifiedExpression
+import org.jetbrains.kotlin.psi.KtSafeQualifiedExpression
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+import org.jetbrains.kotlin.types.TypeUtils
+
+/**
+ * TBC
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: Rule documentation is missing

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * @author Matthew Haughton
+ */
+class UselessCallOnNotNull(config: Config = Config.empty) : Rule(config) {
+    override val issue: Issue = Issue(
+        "UselessCallOnNotNull",
+        Severity.Performance,
+        "This call on not-null type may be reduced or removed. Some calls are intended to be called on nullable " +
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: on a none-nullable type

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                "collection or text types (e.g. String?). When this call is used on a not-null type (e.g. String) " +
+                "it is redundant and will have no effect, so it can be removed.",
+        Debt.FIVE_MINS
+    )
+
+    // Rule adapted from the Kotlin plugin for IntelliJ:
+    // https://github.com/JetBrains/kotlin/blob/f5d0a38629e7d2e7017ee645dc4d4bee60614e93/idea/src/org/jetbrains/kotlin/idea/inspections/collections/UselessCallOnNotNullInspection.kt
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: This should be moved to the rule documentation

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        "kotlin.text.isNullOrEmpty" to Conversion("isEmpty"),
+        "kotlin.text.isNullOrBlank" to Conversion("isBlank")
+    )
+
+    private val uselessNames = toShortFqNames(uselessFqNames.keys)
+
+    @Suppress("ReturnCount")
+    override fun visitQualifiedExpression(expression: KtQualifiedExpression) {
+        super.visitQualifiedExpression(expression)
+        if (bindingContext == BindingContext.EMPTY) return
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: Do we need this check on every usage of the `BindingContext`?

Comment (3flex) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: Not strictly - the `getResolvedCall` call will fail further down with empty `BindingContext`, but checking for it sooner will save a bit of work.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt (hunk 5 above)
Comment (3flex) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: Might go with "non-null reference" as that matches Kotlin docs (https://kotlinlang.org/docs/reference/null-safety.html), what do you think?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt (hunk 7 above)
Comment (arturbosch) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UselessCallOnNotNull.kt]: We could introduce a new rule type which extends its condition check to run only when `bindingContext != BindingContext.EMPTY`?

**Comment by arturbosch** [comment]
Works great!. I've tested this on a fixed mini-project:
```
fun main() {
    val javaList = Java().list.orEmpty()
    val list = listOf(1, 2, 3).orEmpty()
}
```
and it only reports the kotlin list line :)
Could you add such a test and we are ready to merge this :)?

**Comment by 3flex** [comment]
Done!

I didn't add a new rule type as you suggested in your comment, but perhaps that can be added as a later enhancement. There's no interface that the base rule could implement that would ensure the `bindingContext != BindingContext.EMPTY` line was checked for every element visited which seemed like the right way to go, so because the Kotlin compiler doesn't offer that it might get out of date over time. Perhaps you were thinking another way?

**Comment by arturbosch** [comment]
I was thinking of adding a rule type (or just a property for `BaseRule`) where the rule implementor specifies if this rule will only run when `classpath` is set. The `detekt-core` should then take care of running these rules ^^

**Pull request merged**